### PR TITLE
BUG: assign doesnt cast SparseDataFrame to DataFrame

### LINKED ISSUE #19163: to_sparse and assign is giving the wrong results.
Coming from [SO](https://stackoverflow.com/questions/48180206/assigning-a-column-to-a-sparsedataframe), when a boolean scalar is assigned to a new column using `assign`, it is leading to wrong results. Why is that so? 

`df = pd.DataFrame({"a":[1,2,3]}).to_sparse(fill_value=False)`
`df.assign(newcol=False)`

       a  newcol
    0  1     0.0
    1  2     0.0
    2  3     0.0

This works as expected if the dataframe is dense i.e 

   `df.to_dense().assign(newcol=False).to_sparse()`

       a  newcol
    0  1   False
    1  2   False
    2  3   False

Is this a bug or is this a genuine thing that's happening?

### PR DESCRIPTION
**Pull request by hexgnu**
The problem here is that a SparseDataFrame that calls assign should cast
to a DataFrame mainly because SparseDataFrames are a special case.

- [x] closes #19163
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -2713,7 +2713,9 @@ def assign(self, **kwargs):
         8   9  0.549296  2.197225
         9  10 -0.758542  2.302585
         """
-        data = self.copy()
+
+        # See GH19163
+        data = self.copy().to_dense()
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Could you define assign on SparseDataFrame and only density if nescessary?

File: pandas/tests/frame/test_mutate_columns.py [hunk 2]
```diff
@@ -55,6 +55,13 @@ def test_assign(self):
         result = df.assign(A=lambda x: x.A + x.B)
         assert_frame_equal(result, expected)
 
+        # SparseDataFrame
```
Comment (jreback) [inline pandas/tests/frame/test_mutate_columns.py]: make a separate test

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.23.0.txt [hunk 3]
```diff
@@ -448,6 +448,7 @@ Reshaping
 - Bug in :func:`cut` which fails when using readonly arrays (:issue:`18773`)
 - Bug in :func:`Dataframe.pivot_table` which fails when the ``aggfunc`` arg is of type string.  The behavior is now consistent with other methods like ``agg`` and ``apply`` (:issue:`18713`)
 - Bug in :func:`DataFrame.merge` in which merging using ``Index`` objects as vectors raised an Exception (:issue:`19038`)
+- Bug in :func:`DataFrame.assign` which doesn't cast ``SparseDataFrame`` as ``DataFrame``. (:issue:`19163`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: use ``:class`DataFrame` `` and so on here

File: pandas/core/frame.py (hunk 1 above)
Comment (jreback) [reply pandas/core/frame.py]: and actually you don't want to densify, rather you want to do something like (in SparseDataFrame)

```
def assign(self, **kwargs):

      # coerce to a DataFrame
      self = DataFrame(self)
      return self.assign(**kwargs)
```

this actually ends up copying twice though. So the *real* solution is to move  the guts of ``DataFrame.assign``  to ``_assign`` (and leave the copy part in ``.assign``), then call ``._assign`` in the sparse version.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by hexgnu** [comment]
So I updated the PR locally but...

I feel the real problem is inside of SparseArray. 

If you init a SparseArray with False and pass in an index it will assume the dtype is float64 which coerces False to 0.0.

Why does it assume the dtype is float64? Shouldn't it infer it based on infer_dtype_from or something similar?

Thanks!

**Comment by TomAugspurger** [comment]
> If you init a SparseArray with False and pass in an index it will assume the dtype is float64 which coerces False to 0.0.

Can you give an example?

**Comment by hexgnu** [comment]
Sure @TomAugspurger 

```python
In [4]: pd.SparseArray(False, index=[1], fill_value=False)
Out[4]: 
[0.0]
Fill: False
IntIndex
Indices: array([], dtype=int32)
```

The line that coerces it is here:

https://github.com/pandas-dev/pandas/blob/master/pandas/core/sparse/array.py#L198

**Comment by hexgnu** [comment]
Versus

```python
In [5]: pd.SparseArray(False, fill_value=False)
Out[5]: 
[False]
Fill: False
IntIndex
Indices: array([], dtype=int32)
```

**Comment by hexgnu** [comment]
Alright I think I have found the issue. I added some tests and it seems to chooch. Please review and give me any feedback.

Thanks @TomAugspurger and @jreback

File: doc/source/whatsnew/v0.23.0.txt [hunk 4]
```diff
@@ -491,7 +491,7 @@ Groupby/Resample/Rolling
 Sparse
 ^^^^^^
 
--
+- Bug in :class:`SparseArray` where if a scalar and index are passed in it will coerce to float64 regardless of scalar's dtype. (:issue:`19163`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.23.0.txt]: Could you clarify what "passed in" refers to here? Is it specifically `.assign`? Or any method setting / updating the sparse array?

File: pandas/core/sparse/array.py [hunk 5]
```diff
@@ -195,7 +195,7 @@ def __new__(cls, data, sparse_index=None, index=None, kind='integer',
                 data = np.nan
             if not is_scalar(data):
                 raise Exception("must only pass scalars with an index ")
-            values = np.empty(len(index), dtype='float64')
+            values = np.empty(len(index), dtype=infer_dtype_from(data)[0])
```
Comment (TomAugspurger) [inline pandas/core/sparse/array.py]: Is `infer_dtype_from_scalar` more appropriate here, since we've validated that `data` is a scalar?

**Review by TomAugspurger — COMMENTED** [review]
Thanks for tracking this down @hexgnu, just a couple clarifying questions.

And could you add a simple test that hits this directly? Something that constructs a `arr = SparseArray(value, index=[0, 1])` and assert that the `arr.dtype` matches the correct dtype for various `value`.

**Comment by hexgnu** [comment]
@TomAugspurger I did add a test in pandas/tests/sparse/test_array.py that tests this directly instead of indirectly.

File: pandas/core/sparse/array.py (hunk 5 above)
Comment (hexgnu) [reply pandas/core/sparse/array.py]: yea that's a good point. Since infer_dtype_from just checks is_scalar again.

**Comment by hexgnu** [comment]
@jreback @TomAugspurger ping. This is now green after fixing a linting error.

Let me know if you would like me to squash any of these commits.

**Review by TomAugspurger — APPROVED** [review]
Looks good. I'll merge later today, in case @jreback has any comments.

File: pandas/core/sparse/array.py [hunk 6]
```diff
@@ -161,7 +161,8 @@ def __new__(cls, data, sparse_index=None, index=None, kind='integer',
                 data = np.nan
             if not is_scalar(data):
                 raise Exception("must only pass scalars with an index ")
-            values = np.empty(len(index), dtype='float64')
+            values = np.empty(len(index),
```
Comment (jreback) [inline pandas/core/sparse/array.py]: we have a routine for this
construct_1d_from_scalar in pandas.core.dtypes.cast

Comment (jreback) [reply pandas/core/sparse/array.py]: for the empty/fill step

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.23.0.txt [hunk 7]
```diff
@@ -555,6 +555,7 @@ Sparse
 
 - Bug in which creating a ``SparseDataFrame`` from a dense ``Series`` or an unsupported type raised an uncontrolled exception (:issue:`19374`)
 - Bug in :class:`SparseDataFrame.to_csv` causing exception (:issue:`19384`)
+- Bug in constructing a :class:`SparseArray`: if `data` is a scalar and `index` is defined it will coerce to float64 regardless of scalar's dtype. (:issue:`19163`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.23.0.txt]: Whoops, just noticed we don't have `SparseArray` in our API docs so this link won't work. Just

    ``SparseArray``

until we get that sorted out. Also double ticks around data nd index.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @hexgnu

**Comment by jreback** [comment]
great job on sparse fixes! keep em coming!

**Pull request merged**